src/main/java/org/veupathdb/service/eda/ds/plugin/correlationassaymetadata/CorrelationAssayMetadataBipartitenetworkPlugin.java

```diff
@@ -45,19 +57,95 @@ protected boolean computeGeneratesVars() {
   protected void writeResults(OutputStream out, Map<String, InputStream> dataStreams) throws IOException {
-    System.out.println("I'm writing results!");
-    // writeComputeStatsResponseToOutput(out);
+
+    CorrelationStatsResponse stats = getComputeResultStats(CorrelationStatsResponse.class);
+    Number correlationCoefThreshold = getPluginSpec().getCorrelationCoefThreshold() != null ? getPluginSpec().getCorrelationCoefThreshold() : 0.2;
+
+
+    // TEMPORARY: Reshape data using java instead of calling out to R+plot.data
+    // The goal is to transform the response from the correlation app (CorrelationStatsResponse)
+    // into a BipartiteNetwork that can be sent to the frontend. The BipartiteNetwork is composed
+    // of nodes, links, column1NodeIDs, and column2NodeIDs.
+
+    // Prep objects
+    JSONObject bipartiteNetwork = new JSONObject();
+    JSONArray links = new JSONArray();
+    ArrayList<String> nodeIDs = new ArrayList<>();
+    ArrayList<String> column1NodeIDs = new ArrayList<>();
+    ArrayList<String> column2NodeIDs = new ArrayList<>();
+
+    // All the information we need is contained in the stats object. Each row of stats corresponds to a link.
+    // We want to grab the link, the nodes associated with the link, and make note of column assignments for
+    // each node.
+    stats.getStatistics().forEach((correlationRow) -> {
+      
+      // Skip rows that have no correlation coefficient or a correlation coef that is too small. Filtering here prevents us
+      // from showing nodes with no links.
+      if (correlationRow.getCorrelationCoef() == null) return;
+      if (Math.abs(Float.parseFloat(correlationRow.getCorrelationCoef())) < correlationCoefThreshold.floatValue()) return;
+
+      // First add the node ids (data1 and data2 from this row) to our growing list of node ids
+      // We'll worry about duplicates later.
+      nodeIDs.add(correlationRow.getData1());
+      nodeIDs.add(correlationRow.getData2());
+
+      // Add the data1 ids to column 1, and data2 ids to column 2. Again we'll address duplicates later.
+      column1NodeIDs.add(correlationRow.getData1());
+      column2NodeIDs.add(correlationRow.getData2());
+
+      // Next create links
+      // Create source and target objects.
+      JSONObject sourceNode = new JSONObject();
+      sourceNode.put("id", correlationRow.getData1());
+      JSONObject targetNode = new JSONObject();
+      targetNode.put("id", correlationRow.getData2());
+      
+      // Create link with the data from this row and add to links array.
+      JSONObject link = new JSONObject();
+      link.put("source", sourceNode);
+      link.put("target", targetNode);
+      link.put("strokeWidth", correlationRow.getCorrelationCoef());
+      // Link color is the sign of the correlation
+      String color = Float.parseFloat(correlationRow.getCorrelationCoef()) < 0 ? "-1" : "1";
+      link.put("color", color);
+      links.put(link);
+    });
+
+    // Get unique IDs
```

Comment:
you might consider using a Set and converting to a List. im not sure the tradeoff in terms of perf, or if we think there will be enough nodes for that to really matter, etc. dan would probably be the person to ask if you wanted to explore it.

Reply:
lol or just using both, something like `if (hashSet.add(E)) arrayList.add(E);`